src/gap-skipper.js

```diff
@@ -0,0 +1,179 @@
+/**
+ * @file gap-skipper.js
+ */
+import Ranges from './ranges';
+import videojs from 'video.js';
+
+/**
+ * the gap skipper object handles all scenarios
+ * where the player runs into the end of a buffered
+ * region and there is a buffered region ahead.
+ * It then handles the skipping behavior.
+ *
+ * @class GapSkipper
+ */
+export default class GapSkipper {
+  /**
+   * Represents a GapSKipper object.
+   * @constructor
+   * @param {object} options an object that includes the tech and settings
+   */
+  constructor(options) {
+    if (!options.tech.options_.playerId) {
+      return;
+    }
+
+    this.player_ = videojs(options.tech.options_.playerId);
+    this.tech_ = options.tech;
+    this.consecutiveUpdates = 0;
+    this.lastRecordedTime = null;
+    this.timer_ = null;
+
+    if (options.debug) {
+      this.logger_ = videojs.log.bind(videojs, '<gap-skipper>');
+    }
+    this.logger_('<initialize>');
+
+    // The purpose of this function is to emulate the "waiting" event on
+    // browsers that does not emit it when they are stalled waiting for
+    // more data
```

Comment:
[stalled](https://html.spec.whatwg.org/multipage/embedded-content.html#event-media-stalled) is definitely not the right event either. That only applies when the user-agent is also responsible for the fetching of media data. With MSE that responsibility is no longer in it's hands.

There are two main reasons that we can't use `stalled`:
1. What the event means in terms of "how long the data has been not forthcoming" is pretty much left to the browser (though the spec does helpfully say "about 3 seconds").
2. The event can be fired while we are playing. In fact, it often is fired while there is plenty of data in the buffer and the condition usually resolves itself (data arrives) without the browser taking any extra steps.

Waiting is the event IMO that should fire when player has reached the end of the buffer and is waiting for MSE to provide more data.

I agree that we probably shouldn't be firing waiting ourselves. I will stop triggering waiting myself and call the waitingHandler function directly from the work-around for non-waiting-triggering browsers.